### PR TITLE
Preload nodes in lookups

### PR DESCRIPTION
**Pull request by eozturk1**
In `batch_lookups` we utilize node preloading to minimize individual `GET` operations. We can do the same optimization for regular lookups -- which this PR adds. Performance results are shown below. 

In summary, instead of 40 individual `GET` operations, with this PR, we only have 2 `GET` + 8 `BATCH GET` operations which gives a total of 10 compared to 40. This should improve the performance of lookups quite a bit considering database access delays.

---
Without this PR:

```
Post look-up metrics:
Cache hit since last: 47, cached size: 40 items

===================================================
============ Database operation counts ============
===================================================
    SET 2, 
    BATCH SET 3, 
    GET 40, 
    BATCH GET 0
    TOMBSTONE 0
    GET USER STATE 1
    GET USER DATA 0
    GET USER STATE VERSIONS 3
===================================================
============ Database operation timing ============
===================================================
    TIME READ 98 ms
    TIME WRITE 1380 ms
```

With this PR:
```
Post look-up metrics:
Cache hit since last: 80, cached size: 40 items

===================================================
============ Database operation counts ============
===================================================
    SET 2, 
    BATCH SET 3, 
    GET 2, 
    BATCH GET 8
    TOMBSTONE 0
    GET USER STATE 1
    GET USER DATA 0
    GET USER STATE VERSIONS 3
===================================================
============ Database operation timing ============
===================================================
    TIME READ 143 ms
    TIME WRITE 1379 ms
 ```
  
 ---
> **NOTE:** Tests done after patching #320. A single lookup operation costs are shown only for post-lookup. Generated with the diff below:
 
 ```diff
+            mysql_db.flush_cache().await;
+            println!("Pre look-up metrics:");
+            mysql_db.log_metrics(log::Level::Trace).await;
             // Perform 10 random lookup proofs on the published users
             for user in users.iter().choose_multiple(&mut rng, 10) {
                 let key = AkdLabel::from_utf8_str(user);
                 match dir.lookup(key.clone()).await {
                     Err(error) => panic!("Error looking up user information {:?}", error),
                     Ok((proof, root_hash)) => {
+                        println!("Post look-up metrics:");
+                        mysql_db.log_metrics(log::Level::Trace).await;
+                        return;
```
---

### CODE REVIEW

**Comment by slawlor** [comment]
Lints and rustfmt? Change looks good

File: akd/src/append_only_zks.rs [hunk 1]
```diff
@@ -289,6 +291,26 @@ impl Azks {
...
+        let lookup_nodes: Vec<Node> = lookup_infos
+            .iter()
+            .flat_map(|li| vec![li.existent_label, li.marker_label, li.non_existent_label])
+            .map(|l| Node {
+                label: l,
+                hash: EMPTY_DIGEST,
+            })
+            .collect();
+
+        // Load nodes. Note NodeSet will sort these nodes for efficient preloading.
```
Comment (afterdusk) [inline akd/src/append_only_zks.rs:309]: I think we can drop the sorting comment, how `NodeSet` represents the data is an internal detail.

**Review by afterdusk — APPROVED** [review]
LGTM as well 👍

**Comment by kevinlewi** [comment]
Question: should we do something similar like this for history proof generation?

**Comment by eozturk1** [comment]
> Question: should we do something similar like this for history proof generation?

I need to look into how we load the nodes but I think we can. I'll add it next week.

Another possibility is batch history proofs similar to batch lookups.

**Pull request merged**